packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -0,0 +1,574 @@
+    'declare function makeMap<K, V>(): Map<K, V>;',
+    'declare function makeMap<K, V>(): [Map<K, V>];',
+    'declare function arrayOfPairs<T>(): [T, T][];',
+    'declare function fn<T>(input: T): 0 extends 0 ? T : never;',
+    'declare function useFocus<T extends HTMLOrSVGElement>(): [React.RefObject<T>];',
+    `
+      declare function useFocus<T extends HTMLOrSVGElement>(): {
+        ref: React.RefObject<T>;
+      };
+    `,
+    `
+      type Obj = { a: string };
+
+      declare function hasOwnProperty<K extends keyof Obj>(
+        obj: Obj,
+        key: K,
+      ): obj is Obj & { [key in K]-?: Obj[key] };
+    `,
+    `
+      type AsMutable<T extends readonly unknown[]> = {
+        -readonly [Key in keyof T]: T[Key];
+      };
+
+      declare function makeMutable<T>(input: T): MakeMutable<T>;
+    `,
+    `
+      type AsMutable<T extends readonly unknown[]> = {
+        -readonly [Key in keyof T]: T[Key];
+      };
+
+      declare function makeMutable<T>(input: T): MakeMutable<typeof input>;
+    `,
+    `
+      interface Middle {
+        inner: boolean;
+      }
+
+      type Conditional<T extends Middle> = {} & (T['inner'] extends true ? {} : {});
+
+      function withMiddle<T extends Middle = Middle>(options: T): Conditional<T> {
+        return options;
+      }
+    `,
+    `
+      import * as ts from 'typescript';
+
+      declare function forEachReturnStatement<T>(
+        body: ts.Block,
+        visitor: (stmt: ts.ReturnStatement) => T,
+      ): T | undefined;
+    `,
+    `
+      import type { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/types';
+
+      declare const isNodeOfType: <NodeType extends AST_NODE_TYPES>(
+        nodeType: NodeType,
+      ) => node is Extract<TSESTree.Node, { type: NodeType }>;
+    `,
+    `
+      import type { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/types';
+
+      const isNodeOfType =
+        <NodeType extends AST_NODE_TYPES>(nodeType: NodeType) =>
+        (
+          node: TSESTree.Node | null,
+        ): node is Extract<TSESTree.Node, { type: NodeType }> =>
+          node?.type === nodeType;
+    `,
+    `
+      import type { AST_TOKEN_TYPES, TSESTree } from '@typescript-eslint/types';
+
+      export const isNotTokenOfTypeWithConditions =
+        <
+          TokenType extends AST_TOKEN_TYPES,
+          ExtractedToken extends Extract<TSESTree.Token, { type: TokenType }>,
+          Conditions extends Partial<ExtractedToken>,
+        >(
+          tokenType: TokenType,
+          conditions: Conditions,
+        ): ((
+          token: TSESTree.Token | null | undefined,
+        ) => token is Exclude<TSESTree.Token, Conditions & ExtractedToken>) =>
+        (token): token is Exclude<TSESTree.Token, Conditions & ExtractedToken> =>
+          tokenType in conditions;
+    `,
+  ],
+
+  invalid: [
+    {
+      code: 'const func = <T,>(param: T) => null;',
+      errors: [{ messageId: 'sole', data: { name: 'T' } }],
```

Comment:
This is way out of scope for this PR, but has typescript-eslint ever considered switching from unit tests to TypeScript-style baselines? I think a single file containing the source code and any errors in the locations that they were reported would be much easier to read that the giant list of template literals.

Reply:
We have! #56 -> #6499. I personally quite liked the TSLint way of doing things.